lib/Graph/TensorLayout.cpp

```diff
@@ -0,0 +1,557 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <ctype.h>
+#include <memory>
+#include <sstream>
+
+#include "glow/Graph/Graph.h"
+#include "glow/Graph/TensorLayout.h"
+#include "glow/Graph/VerifierHelper.h"
+
+using namespace glow;
+
+/// Checks if two layout descriptions \p lhs and \p rhs describe the same layout
+/// for a value of the type \p ty \returns true if layouts are the same.
+bool glow::checkSameLayout(llvm::StringRef srcLayoutStr,
+                           llvm::StringRef destLayoutStr, TypeRef ty,
+                           const Node *parent, const std::string &prefix,
+                           const TensorLayoutCommon &TLC, bool verbose) {
+  auto srcLayout = TensorLayoutDescription(srcLayoutStr);
+  auto destLayout = TensorLayoutDescription(destLayoutStr);
+  // Are layouts literally the same?
+  if (srcLayout.isSameLayout(destLayout)) {
+    return true;
+  }
+  // Does the type satisfy the dest layout?
+  if (TLC.isSatisfiedBy(ty, destLayout, &srcLayout)) {
+    return true;
+  }
+  if (verbose) {
+    report("\n\n\n");
+    reportContext(parent);
+    report("\n");
+    report(prefix);
+    report("\n");
+    report(parent->getDebugDesc());
+    report("\nMismatching layouts:\n");
+    report("Provided layout\n");
+    report(srcLayout.getDebugDesc());
+    report("\n");
+    report("Expected layout\n");
+    report(destLayout.getDebugDesc());
+    report("\n");
+  }
+  return false;
+}
+
+/// Verifies the correctness of tensor layouts in the function \p F using layout
+/// requirements interface \p TLC.
+bool glow::verifyLayouts(const Function &F, TensorLayoutCommon &TLC,
+                         bool verbose) {
+  bool isValid = true;
+  for (const auto &N : F.getNodes()) {
+    for (unsigned idx = 0, e = N.getNumInputs(); idx < e; ++idx) {
+      auto input = N.getNthInput(idx);
+      auto producerLayout =
+          TLC.getNthResultLayoutRequirements(input.getNode(), input.getResNo());
+      auto consumerLayout = TLC.getNthInputLayoutRequirements(&N, idx);
+      std::string inputName = strFormat("input %d", idx);
+      isValid &= checkSameLayout(producerLayout, consumerLayout,
+                                 input.getType(), &N, inputName, TLC, verbose);
+    }
+  }
+  return isValid;
+}
+
+TensorLayoutDescription::TensorLayoutDescription(std::string text) {
+  if (text.empty()) {
+    // 0-D output
+    numDims_ = 0;
+    return;
+  }
+  parse(text);
+}
+
+static bool isCustomExtension(llvm::StringRef text) {
+  auto nsPos = text.find(':');
+  if (nsPos == llvm::StringRef::npos) {
+    return false;
+  }
+  auto bracketPos = text.find(']');
+  assert(bracketPos != llvm::StringRef::npos && "Expected a closing bracket.");
+  return (bracketPos > nsPos);
+}
+
+// Serialization format -
+// The form for each dimension is as follows:
+// 1. (mandatory) one char representing the current dimension. Either an
```

Comment:
Questions:
- Does it matter of it is upper case or lower case? E.g. in TVM they have different meanings
- Does it need to be a letter?
- Is one character enough for a dimension name?

Reply:
* No it can be either upper or lower, but we can change that if it matters?
* Either an ascii-table english-language letter or a `*` - no emojis :D 
* I assume so? makes it much easier to parse / smaller code, but should be easy to extend that if we change our mind.. I don't see why we'd need more than one.